File: backend/api/app/schemas.py
from pydantic import BaseModel, Field
from typing import List, Optional


class ResearchRequest(BaseModel):
    startup_idea: str = Field(..., min_length=10, description="The startup idea to research")


class ForecastSeries(BaseModel):
    year: int = Field(..., description="Year for the forecast data point")
    value: float = Field(..., description="Forecast value for the year")


class Forecast(BaseModel):
    title: str = Field(..., description="Title of the forecast")
    unit: str = Field(..., description="Unit of measurement (e.g., 'USD Billions', 'Million Users')")
    series: List[ForecastSeries] = Field(..., description="List of forecast data points")
    scenarios: Optional[List[str]] = Field(None, description="Optional list of forecast scenarios")


class ResearchResponse(BaseModel):
    summary: str = Field(..., description="Executive summary of the market research")
    competitors: List[str] = Field(..., description="List of key competitors")
    forecast: Forecast = Field(..., description="Market forecast data")
    sources: List[str] = Field(..., description="List of sources used in the research")

File: frontend/app/lib/schemas.ts
// Zod validation schemas

import { z } from "zod";

export const startupIdeaSchema = z.string()
  .min(10, "Startup idea must be at least 10 characters")
  .max(500, "Startup idea must be less than 500 characters")
  .trim();

export const forecastSeriesSchema = z.object({
  year: z.coerce.number().int(),
  value: z.coerce.number().nonnegative(),
});

export const forecastSchema = z.object({
  title: z.string(),
  unit: z.string(),
  series: z.array(forecastSeriesSchema).min(1, "Forecast must include at least one data point"),
  scenarios: z.array(z.string()).optional(),
});

export const researchResponseSchema = z.object({
  summary: z.string(),
  competitors: z.array(z.string()),
  forecast: forecastSchema,
  sources: z.array(z.string()),
});

// Infer TypeScript types from schemas
export type ForecastSeries = z.infer<typeof forecastSeriesSchema>;
export type Forecast = z.infer<typeof forecastSchema>;
export type ResearchResponse = z.infer<typeof researchResponseSchema>;

